Cython/Compiler/Main.py

```diff
@@ -472,32 +472,66 @@ def create_default_resultobj(compilation_source, options):
     result.embedded_metadata = options.embedded_metadata
     return result
 
-
-def run_pipeline(source, options, full_module_name=None, context=None):
-    from . import Pipeline
-
-    source_ext = os.path.splitext(source)[1]
-    options.configure_language_defaults(source_ext[1:])  # py/pyx
-    if context is None:
-        context = Context.from_options(options)
-
-    # Set up source object
+def setup_source_object(source, source_ext, full_module_name, options, context):
     cwd = os.getcwd()
     abs_path = os.path.abspath(source)
+
     full_module_name = full_module_name or context.extract_module_name(source, options)
     full_module_name = EncodedString(full_module_name)
 
     Utils.raise_error_if_module_name_forbidden(full_module_name)
-
     if options.relative_path_in_code_position_comments:
         rel_path = full_module_name.replace('.', os.sep) + source_ext
         if not abs_path.endswith(rel_path):
             rel_path = source  # safety measure to prevent printing incorrect paths
     else:
         rel_path = abs_path
     source_desc = FileSourceDescriptor(abs_path, rel_path)
-    source = CompilationSource(source_desc, full_module_name, cwd)
+    return CompilationSource(source_desc, full_module_name, cwd)
+
+def run_cached_pipeline(source, options, full_module_name=None, context=None, cache=None, fingerprint=None):
+
+    if context is None:
+        context = Context.from_options(options)
+
+    if cache:
+        if not fingerprint:
+            fingerprint = get_fingerprint(cache, source, options)
```

Comment:
Let's make some arguments non-optional to reduce the overall redundancy and complexity.

I think for a function called `run_cached_pipeline`, we should assume that caching is enabled, so `cache` is not optional. We should also consider whether `fingerprint` really needs to be optional. If we need to pass it in in some cases, why not always pass it in?

I also don't see why `context` should be optional. It just adds more places where we create the context in exactly the same way. Just pass it into both pipeline functions, always.

And, once the context is always created externally, we can also use it to build the `full_module_name` and make that argument non-optional as well.

I would hope that this allows us to distribute the responsibilities more cleanly. For example, I wonder why the Cache isn't responsible to fingerprinting? Couldn't it somehow remember the result for a all source files, so that we don't have to do that externally?